core/src/repair/repair_service.rs

```diff
@@ -677,6 +685,31 @@ impl RepairService {
         }
     }
 
+    pub(crate) fn generate_repairs_for_wen_restart(
+        blockstore: &Blockstore,
+        max_repairs: usize,
+        slots: Arc<RwLock<Vec<Slot>>>,
```

Comment:
I'm not sure of the semantics for when this would be called. Is it necessary to have `Arc<RwLock<Vec<Slot>>>`? Could this just be `&Vec<Slot>`?.

Reply:
This will be a changing list as we receive more RestartLastVotedForkSlots Gossip messages. We don't care about previous results, we only care about the most recent list.

Reply:
Couldn't this just take a copy of the vec so this function doesn't need to be aware of the lock? It seems unnecessary to hold a lock across the blockstore calls. I don't think this logic is currently handling the list changing during the loop anyway.

Reply:
That's fair, changed.

Reply:
> This will be a changing list as we receive more RestartLastVotedForkSlots Gossip messages

What are the semantics of this? Will the service that is parsing gossip and becoming aware of more slots to repair just add new slots to this vector? If so, how will we track progress (ie which slots we have already repaired) to avoid  duplicate repairs?

I'm wondering if a crossbeam channel may be a better fit here. A producer (ie service reading gossip) would fill up the channel, and then the consumer (this code) would read those out. Reading consumes values from the channel which inherently tracks progress

Reply:
The sender which sends the repair list will check the blockstore as well to see which blocks are missing. As more and more Gossip messages come in, it could become aware of more slots to repair, but normally it would be appending at the end of the list, older slots which are in the list will only disappear when they are marked frozen in blockstore.

I originally used a crossbeam channel, but then I realized I didn't care about all the values in the middle, I only care about the newest list, so there is no point keeping all the intermediate versions in the channel.

Maybe I could check whether repairs have been sent for a block on the repair side? This seems like repair internal details which shouldn't be exposed.

Reply:
I'm not sure if it makes sense to have this PR in isolation. Could this just be rolled into the set of changes which will implement setting ` slots_to_repair_for_wen_restart`? Currently this PR is adding an API which is essentially dead code. It would be better to see a more complete change to review the API.

Is `wen_restart` the name we want to use?

Reply:
The whole PR would probably be too big for one or two person to review, you can see the prototype here about how this will be called: https://github.com/solana-labs/solana/pull/33197

I'm always getting complaints about my PR being too big, so I'm now trying to send one tiny piece at a time.

I'm actually not self-aware enough to be happy on the wen_restart name, but my reviewers on https://github.com/solana-foundation/solana-improvement-documents/pull/46 insisted on it. I'd welcome any good suggestions.

Reply:
> The whole PR would probably be too big for one or two person to review, you can see the prototype here about how this will be called: #33197

This change seems pretty manageable to merge with a change consuming this API. I think we have guidelines to avoid intermediate changes which aren't being consumed.

Reply:
> The whole PR would probably be too big for one or two person to review, you can see the prototype here about how this will be called: #33197

This change seems pretty manageable to merge with a change consuming this API. I think we have guidelines to avoid intermediate changes which aren't being consumed.

Reply:
Okay, I'm just hearing the opposite from other reviewers, sigh. I'll close this PR for now and cc you on the bigger PR.